src/solvers/petsc_diff_solver.C

```diff
@@ -350,18 +352,29 @@ void PetscDiffSolver::setup_petsc_data()
       LIBMESH_CHKERR(ierr);
     }
 
+  bool use_petsc_dm = libMesh::on_command_line("--use_petsc_dm");
+
+  // This needs to be called before SNESSetFromOptions
+  if (use_petsc_dm)
+    this->_dm_wrapper.init_and_attach_petscdm(_system, _snes);
+
   ierr = SNESSetFromOptions(_snes);
   LIBMESH_CHKERR(ierr);
 
-  KSP my_ksp;
-  ierr = SNESGetKSP(_snes, &my_ksp);
-  LIBMESH_CHKERR(ierr);
+  // If we're not using PETSc DM, let's keep around
+  // the old style for fieldsplit
+  if (!use_petsc_dm)
+    {
+      KSP my_ksp;
+      ierr = SNESGetKSP(_snes, &my_ksp);
+      LIBMESH_CHKERR(ierr);
 
-  PC my_pc;
-  ierr = KSPGetPC(my_ksp, &my_pc);
-  LIBMESH_CHKERR(ierr);
+      PC my_pc;
+      ierr = KSPGetPC(my_ksp, &my_pc);
```

Comment:
It is not introduced by this PR.But  it would be great if we could get ride of this manual setup of the preconditoner.

Reply:
My intent is to eventually (after several libMesh releases) remove the old `petsc_auto_fieldspilt` in favor of this.

Reply:
In MOOSE, we have our own fieldsplit preconditioner implementation. The preconditioner allows us split the matrix based on not only the physics field but also the geometry. This requires us to implement dm->ops->createfielddecomposition.   

I am guessing you are using the default petsc implementation for this. Then we can not construct  the fieldsplit preconditioner based on the geometry.

Reply:
You are correct, the field split access here is only for physics field split. Again, perhaps we (and whoever is interested) can talk off thread about how you use the domain decomposition so we can integrate that cleanly because, I believe, this would provide all you'd need to for a physics field split and not require anything downstream in MOOSE.